xtensa-lx-rt/src/exception/asm.rs

```diff
@@ -335,356 +150,760 @@ global_asm!(
+        s32e    a11, a0,  -20
+        rfwo
+
+    .section .WindowUnderflow12.text,\"ax\",@progbits
+    .global _WindowUnderflow12
+    .p2align 2
+    .type _WindowUnderflow12,@function
+_WindowUnderflow12:
+        l32e    a0,  a13, -16
+        l32e    a1,  a13, -12
+        l32e    a2,  a13,  -8
+        l32e    a11, a1,  -12
+
+        l32e    a3,  a13,  -4
+        l32e    a4,  a11, -48
+        l32e    a5,  a11, -44
+        l32e    a6,  a11, -40
+        l32e    a7,  a11, -36
+        l32e    a8,  a11, -32
+        l32e    a9,  a11, -28
+        l32e    a10, a11, -24
+        l32e    a11, a11, -20
+        rfwu
+
+    .section .WindowOverflow4.text,\"ax\",@progbits
+    .global _WindowOverflow4
+    .p2align 2
+    .type _WindowOverflow4,@function
+_WindowOverflow4:
+        s32e    a0, a5, -16
+        s32e    a1, a5, -12
+        s32e    a2, a5,  -8
+        s32e    a3, a5,  -4
+        rfwo
+
+    .section .WindowUnderflow4.text,\"ax\",@progbits
+    .global _WindowUnderflow4
+    .p2align 2
+    .type _WindowUnderflow4,@function
+_WindowUnderflow4:
+        l32e    a0, a5, -16
+        l32e    a1, a5, -12
+        l32e    a2, a5,  -8
+        l32e    a3, a5,  -4
+        rfwu
+
+        // inline the _AllocAException saves on the ret.w for WindowUnderflow4
+        // this makes that it just fits, which is needed for the bbci instructions
+
+        .align 4
+        _AllocAException:
+        rsr     a0, WINDOWBASE  // grab WINDOWBASE before rotw changes it
+        rotw    -1              // WINDOWBASE goes to a4, new a0-a3 are scratch
+        rsr     a2, PS
+        extui   a3, a2, 8, 4    // XCHAL_PS_OWB_SHIFT, XCHAL_PS_OWB_BITS
+        xor     a3, a3, a4      // bits changed from old to current windowbase
+        rsr     a4, EXCSAVE1    // restore original a0 (now in a4)
+        slli    a3, a3, 8       // XCHAL_PS_OWB_SHIFT
+        xor     a2, a2, a3      // flip changed bits in old window base
+        wsr     a2, PS          // update PS.OWB to new window base
+        rsync
+
+        bbci    a4, 31, _WindowUnderflow4
+        rotw    -1              // original a0 goes to a8
+        bbci    a8, 30, _WindowUnderflow8
+        rotw    -1
+        j               _WindowUnderflow12
+
+    .section .KernelExceptionVector.text,\"ax\",@progbits
+    .global _KernelExceptionVector
+    .p2align 2
+    .type _KernelExceptionVector,@function
+_KernelExceptionVector:
+        wsr a0, EXCSAVE1 // preserve a0
+        rsr a0, EXCCAUSE // get exception cause
+
+        beqi a0, 5, .AllocAException
+
+        call0 __naked_kernel_exception
+
+    .section .UserExceptionVector.text,\"ax\",@progbits
+    .global _UserExceptionVector
+    .p2align 2
+    .type _UserExceptionVector,@function
+_UserExceptionVector:
+        wsr a0, EXCSAVE1 // preserve a0
+        rsr a0, EXCCAUSE // get exception cause
+
+        beqi a0, 5, .AllocAException
+
+        call0 __naked_user_exception
+
+.AllocAException:
+        call0  _AllocAException
```

Comment:
Can we save this call0 by just jumping to _AllocAException?

Reply:
That should be a very cold code path (I'm not even sure we can exercise it currently), so I'd say I'm not going to change it in this PR. Feel free to try it out in another PR though.